[RFC] terminal : don't set vterm size to 0 (workaround #2732)

frederikvs opened the pull request:
I've found a workaround for issue #2732. Currently, when resizing, currently we call `vterm_set_size` 3 times. First to size 0, then to the size of one of the windows, then to the size of the other.

It looks like libvterm doesn't handle a terminal of height and/or width 0 properly. And to be honest that doesn't really make a lot of sense. Just checking for that solves the most immediate problem.

Currently some tests fail since they were affected by the bug. I'll report back when I've updated those tests.

frederikvs committed:
terminal : don't set vterm size to 0 (workaround #2732)

Pyrohh commented:
Thanks for looking into this, but why not send a patch to libvterm? They're quite responsive from my experience.

frederikvs commented:
I haven't had the time yet to look into the underlying libvterm-problem. I just noticed that we could prevent the problem by not telling libvterm that its size is 0.
And to me that makes sense, that your terminal's size should never be set to 0. It's just an intermediate state, and neovim sets the size to the intended value immediately after.

splinterofchaos commented on src/nvim/terminal.c:
```diff
@@ -313,6 +313,10 @@ void terminal_resize(Terminal *term, uint16_t width, uint16_t height)
 
+  if (height == 0 || width == 0) {
+    return;
+  }
```
Should this maybe be above the `FOR_ALL_WINDOWS_IN_TAB` block? Or maybe we should be checking this there to make sure that `MIN` isn't the problem.

frederikvs replied:
Above the FOR_ALL_WINDOWS_IN_TAB this wouldn't work yet. If 0 was passed as parameter, the width or height would be set to the current value (not 0, so no return), and then we'd find the minimum (again 0, causing the same problem). 
Also, as you mentioned in a comment to PR #2683, having width & height as a parameter doesn't really add value, since we need to walk across all windows anyway. There are some other problems with this function as well - currently the terminal's size can only decrease, never increase. All this and more is addressed in PR #2683, which can be completed when this PR gets merged in.

The current PR is meant as a minimal bugfix that would enable me to make greater improvements in the other one.

splinterofchaos replied:
I see. Just for the record, normally, I'd request that such changes be put in the main PR, but I do prefer to get bug fixes merged sooner than later, and it's only one commit.

However, I still have to ask why not check for this in the above loop? It doesn't seem like `height` or `width` should be set to zero under any reasonable circumstances. Could `curwidth` have been zero initially? Can `wp->w_width` ever be zero? Does subtracting `win_col_off()` cause it? I worry that this might hide a bug rather than fix it (even if only temporarily).

Ideally, the commit message should document an analysis of the problem that justifies the proposed fix.

frederikvs replied:
> Just for the record, normally, I'd request that such changes be put in the main PR, but I do prefer to get bug fixes merged sooner than later, and it's only one commit.

In the other PR there's some more debate going on, so I didn't want to add yet another commit to it, since it's a bugfix. Throwing those together would probably delay both the bugfix and the new behaviour I'd like to introduce. I was hoping to get this merged in rather quickly.

As for analysis of the problem : 

The reason we end up with a 0 height is the following : when splitting a window (`win_split_ins` in `window.c`), we first set the new width (`win_new_width`) which updates the terminal via `terminal_resize` (but at that time the height hasn't been set yet). After that we set the new height (`win_new_height`) for the one window, then for the other. This function also calls `terminal_resize`, with more correct values.

With a breakpoint at `terminal_resize`, doing a `:sp` and each time printing the stack and the arguments I get the following : 

```
Breakpoint 1, terminal_resize (term=0x7fcd28b0a000, width=80, height=height@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/terminal.c:286
286	{
(gdb) where
#0  terminal_resize (term=0x7fcd28b0a000, width=80, height=height@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/terminal.c:286
#1  0x000000000047ebfd in win_new_width (wp=wp@entry=0x7fcd28925000, 
    width=<optimized out>) at /home/frederik/neovim/neovim/src/nvim/window.c:4748
#2  0x0000000000482ce8 in win_split_ins (size=size@entry=0, flags=flags@entry=0, 
    new_wp=new_wp@entry=0x0, dir=dir@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/window.c:894
#3  0x00000000004831ad in win_split (size=0, flags=0)
    at /home/frederik/neovim/neovim/src/nvim/window.c:535
#4  0x00000000004919cb in ex_splitview (eap=0x7fffbdaac6f0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:6265
#5  0x000000000048fb30 in do_one_cmd (cmdlinep=cmdlinep@entry=0x7fffbdaac840, 
    sourcing=sourcing@entry=0, cstack=cstack@entry=0x7fffbdaac8d0, 
    fgetline=fgetline@entry=0x4ba7e5 <getexline>, cookie=cookie@entry=0x0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:2196
#6  0x0000000000490492 in do_cmdline (cmdline=<optimized out>, cmdline@entry=0x0, 
    fgetline=0x4ba7e5 <getexline>, cookie=cookie@entry=0x0, flags=0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:622
#7  0x000000000049bcc8 in nv_colon (cap=0x7fffbdaace00)
    at /home/frederik/neovim/neovim/src/nvim/normal.c:4078
#8  0x00000000004a42ac in normal_cmd (oap=oap@entry=0x7fffbdaacea0, 
    toplevel=toplevel@entry=true) at /home/frederik/neovim/neovim/src/nvim/normal.c:911
#9  0x00000000004f600e in main_loop (cmdwin=cmdwin@entry=0, noexmode=noexmode@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/main.c:706
#10 0x00000000004f6f32 in main (argc=<optimized out>, argv=<optimized out>)
    at /home/frederik/neovim/neovim/src/nvim/main.c:493
(gdb) info args
term = 0x7fcd28b0a000
width = 80
height = 0
(gdb) c
Continuing.

Breakpoint 1, terminal_resize (term=0x7fcd28b0a000, width=width@entry=0, height=11)
    at /home/frederik/neovim/neovim/src/nvim/terminal.c:286
286	{
(gdb) where
#0  terminal_resize (term=0x7fcd28b0a000, width=width@entry=0, height=11)
    at /home/frederik/neovim/neovim/src/nvim/terminal.c:286
#1  0x000000000047f11f in win_new_height (wp=wp@entry=0x7fcd28925000, 
    height=height@entry=11) at /home/frederik/neovim/neovim/src/nvim/window.c:4726
#2  0x0000000000482d09 in win_split_ins (size=size@entry=0, flags=flags@entry=0, 
    new_wp=new_wp@entry=0x0, dir=dir@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/window.c:901
#3  0x00000000004831ad in win_split (size=0, flags=0)
    at /home/frederik/neovim/neovim/src/nvim/window.c:535
#4  0x00000000004919cb in ex_splitview (eap=0x7fffbdaac6f0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:6265
#5  0x000000000048fb30 in do_one_cmd (cmdlinep=cmdlinep@entry=0x7fffbdaac840, 
    sourcing=sourcing@entry=0, cstack=cstack@entry=0x7fffbdaac8d0, 
    fgetline=fgetline@entry=0x4ba7e5 <getexline>, cookie=cookie@entry=0x0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:2196
#6  0x0000000000490492 in do_cmdline (cmdline=<optimized out>, cmdline@entry=0x0, 
    fgetline=0x4ba7e5 <getexline>, cookie=cookie@entry=0x0, flags=0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:622
#7  0x000000000049bcc8 in nv_colon (cap=0x7fffbdaace00)
    at /home/frederik/neovim/neovim/src/nvim/normal.c:4078
#8  0x00000000004a42ac in normal_cmd (oap=oap@entry=0x7fffbdaacea0, 
    toplevel=toplevel@entry=true) at /home/frederik/neovim/neovim/src/nvim/normal.c:911
#9  0x00000000004f600e in main_loop (cmdwin=cmdwin@entry=0, noexmode=noexmode@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/main.c:706
#10 0x00000000004f6f32 in main (argc=<optimized out>, argv=<optimized out>)
    at /home/frederik/neovim/neovim/src/nvim/main.c:493
(gdb) info args
term = 0x7fcd28b0a000
width = 0
height = 11
(gdb) c    
Continuing.

Breakpoint 1, terminal_resize (term=0x7fcd28b0a000, width=width@entry=0, height=10)
    at /home/frederik/neovim/neovim/src/nvim/terminal.c:286
286	{
(gdb) where
#0  terminal_resize (term=0x7fcd28b0a000, width=width@entry=0, height=10)
    at /home/frederik/neovim/neovim/src/nvim/terminal.c:286
#1  0x000000000047f11f in win_new_height (wp=wp@entry=0x7fcd2887e000, 
    height=height@entry=10) at /home/frederik/neovim/neovim/src/nvim/window.c:4726
#2  0x0000000000482d3c in win_split_ins (size=size@entry=0, flags=flags@entry=0, 
    new_wp=new_wp@entry=0x0, dir=dir@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/window.c:906
#3  0x00000000004831ad in win_split (size=0, flags=0)
    at /home/frederik/neovim/neovim/src/nvim/window.c:535
#4  0x00000000004919cb in ex_splitview (eap=0x7fffbdaac6f0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:6265
#5  0x000000000048fb30 in do_one_cmd (cmdlinep=cmdlinep@entry=0x7fffbdaac840, 
    sourcing=sourcing@entry=0, cstack=cstack@entry=0x7fffbdaac8d0, 
    fgetline=fgetline@entry=0x4ba7e5 <getexline>, cookie=cookie@entry=0x0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:2196
#6  0x0000000000490492 in do_cmdline (cmdline=<optimized out>, cmdline@entry=0x0, 
    fgetline=0x4ba7e5 <getexline>, cookie=cookie@entry=0x0, flags=0)
    at /home/frederik/neovim/neovim/src/nvim/ex_docmd.c:622
#7  0x000000000049bcc8 in nv_colon (cap=0x7fffbdaace00)
    at /home/frederik/neovim/neovim/src/nvim/normal.c:4078
#8  0x00000000004a42ac in normal_cmd (oap=oap@entry=0x7fffbdaacea0, 
    toplevel=toplevel@entry=true) at /home/frederik/neovim/neovim/src/nvim/normal.c:911
#9  0x00000000004f600e in main_loop (cmdwin=cmdwin@entry=0, noexmode=noexmode@entry=0)
    at /home/frederik/neovim/neovim/src/nvim/main.c:706
#10 0x00000000004f6f32 in main (argc=<optimized out>, argv=<optimized out>)
    at /home/frederik/neovim/neovim/src/nvim/main.c:493
(gdb) info args
term = 0x7fcd28b0a000
width = 0
height = 10
```
If you like, I can add this explanation to the commit message.

Concerning doing it in the loop, is it something like this you have in mind? 
```
void terminal_resize(Terminal *term, uint16_t width, uint16_t height)
{
  if (term->closed) {
    // will be called after exited if two windows display the same terminal and
    // one of the is closed as a consequence of pressing a key.
    return;
  }
  int curwidth, curheight;
  uint16_t width_temp, height_temp;
  vterm_get_size(term->vt, &curheight, &curwidth);

  if (!width) {
    width = (uint16_t)curwidth;
  }

  if (!height) {
    height = (uint16_t)curheight;
  }

  // The new width/height are the minimum for all windows that display the
  // terminal in the current tab.
  FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
    if (!wp->w_closing && wp->w_buffer == term->buf) {
      width_temp = (uint16_t)MIN(width,
          (uint16_t)(wp->w_width - win_col_off(wp)));
      if (width_temp != 0) {
        width = width_temp;
      }
      height_temp = (uint16_t)MIN(height, (uint16_t)wp->w_height);
      if (height_temp != 0) {
        height = height_temp;
      }
    }
  }

  if (curheight == height && curwidth == width) {
    return;
  }

  vterm_set_size(term->vt, height, width);
  vterm_screen_flush_damage(term->vts);
  term->pending_resize = true;
  invalidate_terminal(term, -1, -1);
}
```
I'm not really sure it makes all that much difference, either fix is fine by me :-)

splinterofchaos replied:
> when splitting a window (win_split_ins in window.c), we first set the new width (win_new_width) which updates the terminal via terminal_resize (but at that time the height hasn't been set yet). After that we set the new height (win_new_height) for the one window, then for the other. This function also calls terminal_resize, with more correct values.

I see. Now I understand why exiting when `width` or `height` equals zero fixes the problem--we called `terminal_resize` before actually knowing what size it should be. I think the new window's dimensions (from `win_split() -> win_alloc()`) have been initialized to zero. If that's the case, it would be more optimal to have something like this in the `FOR_ALL_WINDOWS_IN_TAB` block:
```c
if (wp->w_height == 0) {
  return;  // Window still being initialized. win_new_height() will call this again.
}
```
We only need to check `w_height` `win_alloc()` sets `w_width` to `Rows`.

Sorry I'm delaying this so much. I just want to make sure we're fixing it in the right way.

frederikvs replied:
Maybe we should fix this on two fronts : 
  * window.c should not call terminal_resize when it isn't initialized completely yet
  * terminal.c should not let a height or width of 0 pass to vterm_set_size

The second case shouldn't happen, but it's a safeguard for if/when something would change in window.c. (anything with a 400-line function is a problem waiting to happen)

I would still call `win_new_width` from `win_split_ins`, since that does a bunch of other things too. But `win_new_width` could do the check for zero-height

commit added.

splinterofchaos replied:
:+1:

justinmk commented:
Nice! Looks like this may have fixed the OS X indeterminism on travis (except for test61 which may be a different issue).

frederikvs committed:
fix for #2732 : win_new_width checks height before terminal_resize

When splitting the window (win_split_ins), function win_new_width is
already called before the height has been set. This calls
terminal_resize, which passes a height of 0 on to libvterm, which
doesn't handle a height of 0 properly.
A fix is already in place in terminal.c for not passing on the height,
but strictly speaking, it doesn't make sense for window to call
terminal_resize when it isn't initialized completely yet.

splinterofchaos commented:
LGTM :+1:  -- great work!

frederikvs commented:
Great :-)
Anything else that is expected of me before this can be merged?

justinmk commented:
I'm going to make a test PR that is based on this one, with test61 removed, to see if any other tests fail.

justinmk commented:
Merged. Thank you!

justinmk commented on src/nvim/window.c:
```diff
@@ -4745,7 +4745,9 @@ void win_new_width(win_T *wp, int width)
   if (wp->w_buffer->terminal) {
-    terminal_resize(wp->w_buffer->terminal, wp->w_width, 0);
+    if (wp->w_height != 0) {
+      terminal_resize(wp->w_buffer->terminal, wp->w_width, 0);
```
Checking `wp->w_height` seems to add some technical debt. We're baking-in knowledge of `terminal_resize `'s implementation (we're passing the literal `0` yet we're checking `wp->w_height` for non-zero... that makes no sense). I suggest removing this check in your other PR.

frederikvs replied:
Actually those parameters to terminal_resize are useless. Terminal_resize will loop over all windows on the terminal-buffer to figure out what the minimum size is, and use that as the new size. Passing in the value for this specific window adds no value.

The other PR already contains a commit which removes those unnecessary parameters.

justinmk replied:
:+1:
